File: apps/api/signalyze_api/engine/signals/date_entropy.py
"""Twin of ``signals/date-entropy.ts``."""

from __future__ import annotations

import math
from collections.abc import Sequence

from signalyze_api.engine.data_files import THRESHOLDS
from signalyze_api.engine.mathutil import entropy_bits, months_between_inclusive, ramp, round6
from signalyze_api.engine.prepare import PreparedReview
from signalyze_api.engine.types import unavailable
from signalyze_api.models import SignalResult

T = THRESHOLDS.date_entropy


def date_entropy(reviews: Sequence[PreparedReview]) -> SignalResult:
    """Normalised Shannon entropy of per-month counts over the first..last month span.

    value = normalised entropy (1 = even). Unusualness grows as the distribution
    concentrates (1 - entropy).
    """
    counts: dict[str, int] = {}
    first = ""
    last = ""
    for r in reviews:
        counts[r.month] = counts.get(r.month, 0) + 1
        if first == "" or r.month < first:
            first = r.month
        if last == "" or r.month > last:
            last = r.month
    months_in_span = months_between_inclusive(first, last)
    if months_in_span < T.minMonths:
        return unavailable(
            "date_entropy", {"monthsInSpan": months_in_span, "minMonths": T.minMonths}
        )
    values = list(counts.values())
    h = entropy_bits(values)
    normalized = h / math.log2(months_in_span)
    busiest = ""
    busiest_count = 0
    for month, c in sorted(counts.items(), key=lambda item: item[0]):
        if c > busiest_count:
            busiest = month
            busiest_count = c
    return SignalResult(
        id="date_entropy",
        unusualness=round6(ramp(1 - normalized, T.low, T.high)),
        value=round6(normalized),
        details={
            "normalizedEntropy": round6(normalized),
            "monthsInSpan": months_in_span,
            "activeMonths": len(counts),
            "busiestMonth": busiest,
            "busiestMonthShare": round6(busiest_count / len(reviews)),
        },
        available=True,
    )

File: packages/signals/src/signals/date-entropy.ts
import type { SignalResult } from '@signalyze/shared';
import thresholds from '../../data/thresholds.json';
import { entropyBits, monthsBetweenInclusive, ramp, round6 } from '../math';
import type { PreparedReview } from '../prepare';
import { unavailable } from '../types';

const T = thresholds.date_entropy;

/**
 * Normalised Shannon entropy of the per-month review counts over the span from
 * the first to the last review month. value = normalised entropy (1 = even).
 * Unusualness grows as the distribution concentrates (1 - entropy).
 */
export function dateEntropy(reviews: readonly PreparedReview[]): SignalResult {
  const counts = new Map<string, number>();
  let first = '';
  let last = '';
  for (const r of reviews) {
    counts.set(r.month, (counts.get(r.month) ?? 0) + 1);
    if (first === '' || r.month < first) first = r.month;
    if (last === '' || r.month > last) last = r.month;
  }
  const monthsInSpan = monthsBetweenInclusive(first, last);
  if (monthsInSpan < T.minMonths) {
    return unavailable('date_entropy', { monthsInSpan, minMonths: T.minMonths });
  }
  const values = [...counts.values()];
  const h = entropyBits(values);
  const normalized = h / Math.log2(monthsInSpan);
  let busiest = '';
  let busiestCount = 0;
  for (const [month, c] of [...counts.entries()].sort((a, b) => (a[0] < b[0] ? -1 : 1))) {
    if (c > busiestCount) {
      busiest = month;
      busiestCount = c;
    }
  }
  return {
    id: 'date_entropy',
    unusualness: round6(ramp(1 - normalized, T.low, T.high)),
    value: round6(normalized),
    details: {
      normalizedEntropy: round6(normalized),
      monthsInSpan,
      activeMonths: counts.size,
      busiestMonth: busiest,
      busiestMonthShare: round6(busiestCount / reviews.length),
    },
    available: true,
  };
}
